### ISSUE TITLE
docker-enter/nsenter complains about tty process group

### ISSUE THREAD

**Issue by ghost** [YungSang/boot2docker-vagrant-box#9]
There seems to be an issue with either `docker-enter` or `nsenter` as bundled with the current box. Below is an example session that demonstrates the problem.

See the line just after exiting from the first `nsenter` usage:

```
-sh: 1: Cannot set tty process group (No such process)
```

This doesn't happen with every container but does reliably happen with the following script.

Using `docker-enter` and `nsenter` from [jpetazzo/nsenter](http://github.com/jpetazzo/nsenter) seems to make the message go away. It looks like his `docker-enter` implementation is a little different, clearing the environment and starting a login shell. The `nsenter` he provides is also more recent.

```
                        ##        .
                  ## ## ##       ==
               ## ## ## ##      ===
           /""""""""""""""""\___/ ===
      ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~
           \______ o          __/
             \    \        __/
              \____\______/
 _                 _   ____     _            _
| |__   ___   ___ | |_|___ \ __| | ___   ___| | _____ _ __
| '_ \ / _ \ / _ \| __| __) / _` |/ _ \ / __| |/ / _ \ '__|
| |_) | (_) | (_) | |_ / __/ (_| | (_) | (__|   <  __/ |
|_.__/ \___/ \___/ \__|_____\__,_|\___/ \___|_|\_\___|_|
boot2docker: 1.1.2
             nfs+nsenter : e00f8a9 - Wed Jul 23 22:09:44 UTC 2014
docker@boot2docker:~$ docker run -it --name='ghci' darinmorrison/haskell bash -cl ghci
GHCi, version 7.8.3: http://www.haskell.org/ghc/  :? for help
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.
Prelude> sum [0 .. 100]
5050
Prelude> docker@boot2docker:~$
docker@boot2docker:~$ nsenter --version
nsenter from util-linux 2.23.1
docker@boot2docker:~$ docker-enter ghci
nsenter --target 1510 --mount --uts --ipc --net --pid --
# exit
-sh: 1: Cannot set tty process group (No such process)
docker@boot2docker:~$ rm -f /usr/local/bin/docker-enter
docker@boot2docker:~$ rm -f /usr/local/bin/nsenter
docker@boot2docker:~$ docker run --rm -v /usr/local/bin:/target jpetazzo/nsenter
Installing nsenter to /target
Installing docker-enter to /target
docker@boot2docker:~$ nsenter --version
nsenter from util-linux 2.24
docker@boot2docker:~$ docker-enter ghci
nsenter: cannot open /proc/1510/ns/ipc: Permission denied
docker@boot2docker:~$ sudo docker-enter ghci
root@5fb557898b90:~# exit
logout
docker@boot2docker:~$ exit
Connection to 127.0.0.1 closed.
```

**Comment by YungSang** [comment]
@darinmorrison Thank you for the information.
I will try it.

**Comment by YungSang** [comment]
`nsenter` uses `$SHELL` as a login shell into a container. It is from the specification of `nsenter`. 

http://man7.org/linux/man-pages/man1/nsenter.1.html

> If program is not given, then ``${SHELL}'' is run (default: /bin/sh).

In this case, `nsenter` tries to use `/bin/sh` in the container.
This seems a problem for `darinmorrison/haskell`.

```
docker@boot2docker:~$ echo $SHELL
/bin/sh
docker@boot2docker:~$ docker-enter ghci
nsenter --target 1828 --mount --uts --ipc --net --pid --
# exit
-sh: 1: Cannot set tty process group (No such process)
docker@boot2docker:~$ docker-enter ghci bash
nsenter --target 1828 --mount --uts --ipc --net --pid -- bash
root@1a8ecb18d439:/# exit
exit
docker@boot2docker:~$
```

`docker-enter` from `jpetazzo/nsenter` uses `su - root` instead of `/bin/sh`, so it's going to use `bash`.
That's why it looks good.

On the other hand, with `nsenter` from `jpetazzo/nsenter`,  the same issue happens.

```
docker@boot2docker:~$ sudo nsenter --target 1828 --mount --uts --ipc --net --pid --
# exit
-sh: 1: Cannot set tty process group (No such process)
docker@boot2docker:~$ sudo nsenter --target 1828 --mount --uts --ipc --net --pid -- env --ignore-environment -- sh
# exit
sh: 1: Cannot set tty process group (No such process)
docker@boot2docker:~$
```

In conclusion, it doesn't depend on versions of `nsenter`, but `docker-enter`.
However, I still prefer to use my `docker-enter` in order to follow the specification of `nsenter`.

Please use `bash` as a shell for `darinmorrison/haskell`.

**Comment by ghost** [comment]
Ah, okay.

I did notice that `docker-enter` from `jpetazzo/nsenter` used `su - root` to start the login shell and thought I tried comparing both versions with the same command too but guess I must have missed it.

Thanks for tracking this down :]

**Comment by YungSang** [comment]
@darinmorrison You're welcome. Thank you for letting me know this.
